src/kmp.h

```diff
@@ -0,0 +1,68 @@
+#ifndef KMP_H
+#define KMP_H
+
+#include <malloc.h>
+
+int *compute_border_array(char *pattern, int pattern_len) {
+    if (pattern_len == 0) return NULL;
+    int *ba = malloc(sizeof(int)*pattern_len);
+    ba[0] = 0;
+    for (int i = 1; i < pattern_len; i++) {
+        // Test if we can extend border. Otherwise, reset.
+        if (pattern[i] == pattern[ba[i-1]]) {
+            ba[i] = ba[i-1]+1;
+        } else {
+            ba[i] = 0;
+        }
+    }
+    return ba;
+}
+
+int *compute_restricted_border_array(char *pattern, int pattern_len) {
+    // Compute and change the border array.
+    // Could be done faster (and uglier), but we assume pattern_len is not long.
+    int *bax = compute_border_array(pattern, pattern_len);
+    for (int i = 1; i < pattern_len; i++) {
+        if (pattern[i+1] == pattern[bax[i]]) {
+            // Next char after border must be different
+            bax[i] = bax[bax[i]-1];
+        }
+    }
+    return bax;
+}
+
+// Checks sequence[seq_index:] for pattern. Returns index when exact pattern is found. Otherwise -1.
+int kmp_get_position_of_next_match(char *sequence, char *pattern, int seq_index, int pattern_len, int *bax) {
+    // We define, that the empty string can be found everywhere
+    if (pattern_len == 0) return seq_index+1;
+    int seq_len = (int) strlen(sequence);
+
+    int i = seq_index;
+    int j = 0;
```

Comment:
Be careful here! When you set `j` to zero, you also set the current border matched to zero. You are, in effect, restarting the search at this index, but presumably, you plan to ask for multiple matches. If you ask for the next match from one past the previous match, you shouldn't reset the border.

Assume that you are searching for $p=a^m$ in $x=a^n$:

```
x = aaaaaaaaaaaaaaaaaaaa
p = aaaa
       ^ - match at 0
```

If you move `p` to position 1 now

```
x = aaaaaaaaaaaaaaaaaaaa
p =  aaaa
     ... -- these already match
```

you shouldn't start matching from the beginning of `p`. You already know that you match the first three `a` because you just did that.

It is because you keep moving along from where you already matched that the algorithm runs in linear time. If we restarted from 1 in the example above, and then 2 and 3 and so on, then we were back to $O(nm)$.

I don't know how you use the function yet, I haven't got that far, but this is dangerous waters.